server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java

```diff
@@ -189,6 +200,92 @@ void testGetClientAssertionCustomSingingKeyButNoCertificate() throws ParseExcept
+    // Then
+    assertFalse(jwtClientAuthentication.validateClientJwt(getMockedRequestParameter("test", INVALID_CLIENT_JWT), getMockedClientJwtConfiguration(null), "identity"));
+  }
+
+  @Test
+  void testWrongAssertionInvalidateClientId() throws Exception {
+    // Given
+    jwtClientAuthentication = new JwtClientAuthentication(keyInfoService, oidcMetadataFetcher);
+    // Then
+    Exception exception = assertThrows(BadCredentialsException.class, () ->
+        jwtClientAuthentication.validateClientJwt(getMockedRequestParameter(null,  jwtClientAuthentication.getClientAssertion(config)),
+            getMockedClientJwtConfiguration(null), "test"));
+    assertEquals("Wrong client_assertion", exception.getMessage());
+  }
+
+  @Test
+  void testBadAlgorithmInvalidateClientId() throws Exception {
+    // Given
+    jwtClientAuthentication = new JwtClientAuthentication(keyInfoService, oidcMetadataFetcher);
+    ClientJwtConfiguration clientJwtConfiguration = getMockedClientJwtConfiguration(null);
+    when(oidcMetadataFetcher.fetchWebKeySet(clientJwtConfiguration)).thenReturn(clientJwtConfiguration.getJwkSet());
+    // Then
+    Exception exception = assertThrows(BadCredentialsException.class, () ->
+        jwtClientAuthentication.validateClientJwt(getMockedRequestParameter(null, INVALID_CLIENT_ALG),
+            clientJwtConfiguration, "identity"));
+    assertEquals("Bad client_assertion algorithm", exception.getMessage());
+  }
+
+  @Test
+  void testOidcFetchFailed() throws Exception {
+    // Given
+    jwtClientAuthentication = new JwtClientAuthentication(keyInfoService, oidcMetadataFetcher);
+    ClientJwtConfiguration clientJwtConfiguration = getMockedClientJwtConfiguration(null);
+    when(oidcMetadataFetcher.fetchWebKeySet(clientJwtConfiguration)).thenThrow(new OidcMetadataFetchingException(""));
+    // Then
+    Exception exception = assertThrows(BadCredentialsException.class, () ->
+        jwtClientAuthentication.validateClientJwt(getMockedRequestParameter(null, jwtClientAuthentication.getClientAssertion(config)),
+            clientJwtConfiguration, "identity"));
+    assertEquals("Bad client_assertion", exception.getMessage());
+  }
+
+  @Test
+  void testInvalidateClientAssertion() throws Exception {
+    // Given
+    jwtClientAuthentication = new JwtClientAuthentication(keyInfoService, oidcMetadataFetcher);
+    ClientJwtConfiguration clientJwtConfiguration = getMockedClientJwtConfiguration("extId");
+    when(oidcMetadataFetcher.fetchWebKeySet(clientJwtConfiguration)).thenReturn(clientJwtConfiguration.getJwkSet());
+    // When
+    String clientAssertion = (String) jwtClientAuthentication.getClientAssertion(config);
+    // Then
+    Exception exception = assertThrows(BadCredentialsException.class, () ->
+        jwtClientAuthentication.validateClientJwt(getMockedRequestParameter(null, clientAssertion), clientJwtConfiguration, "identity"));
+    assertEquals("Untrusted client_assertion", exception.getMessage());
```

Comment:
Please comment here how the expected exception is caused

Reply:
done